latch_cli/services/cp/upload.py

```diff
@@ -299,70 +287,94 @@ class StartUploadReturnType:
     dest: str
 
 
+MAX_RETRIES = 5
+
+
 def start_upload(
     src: Path,
     dest: str,
     progress_bars: Optional[ProgressBars] = None,
+    throttle: Optional[Throttle] = None,
+    latency_q: Optional["LatencyQueueType"] = None,
 ) -> Optional[StartUploadReturnType]:
-    if not src.exists():
-        raise ValueError(f"Could not find {src}: no such file or link")
-
-    if src.is_symlink():
-        src = src.resolve()
-
-    content_type, _ = mimetypes.guess_type(src)
-    if content_type is None:
-        with open(src, "rb") as f:
-            sample = f.read(units.KiB)
-
-        try:
-            sample.decode()
-            content_type = "text/plain"
-        except UnicodeDecodeError:
-            content_type = "application/octet-stream"
-
-    file_size = src.stat().st_size
-    if file_size > latch_constants.maximum_upload_size:
-        raise ValueError(
-            f"File is {with_si_suffix(file_size)} which exceeds the maximum upload size"
-            " (5TiB)"
+    try:
+        if not src.exists():
+            raise ValueError(f"Could not find {src}: no such file or link")
+
+        if src.is_symlink():
+            src = src.resolve()
+
+        content_type, _ = mimetypes.guess_type(src)
+        if content_type is None:
+            with open(src, "rb") as f:
+                sample = f.read(units.KiB)
+
+            try:
+                sample.decode()
+                content_type = "text/plain"
+            except UnicodeDecodeError:
+                content_type = "application/octet-stream"
+
+        file_size = src.stat().st_size
+        if file_size > latch_constants.maximum_upload_size:
+            raise ValueError(
+                f"File is {with_si_suffix(file_size)} which exceeds the maximum upload"
+                " size (5TiB)"
+            )
+
+        part_count = min(
+            latch_constants.maximum_upload_parts,
+            math.ceil(file_size / latch_constants.file_chunk_size),
+        )
+        part_size = max(
+            latch_constants.file_chunk_size,
+            math.ceil(file_size / latch_constants.maximum_upload_parts),
         )
 
-    part_count = min(
-        latch_constants.maximum_upload_parts,
-        math.ceil(file_size / latch_constants.file_chunk_size),
-    )
-    part_size = max(
-        latch_constants.file_chunk_size,
-        math.ceil(file_size / latch_constants.maximum_upload_parts),
-    )
+        retries = 0
+        while retries < MAX_RETRIES:
+            if throttle is not None:
+                time.sleep(throttle.get_delay())
```

Comment:
From what I can see here, a delay is set and all the start uploads wait. Then when the delay is unset, they all wake up and spam together. I think some jitter on the wakeup would help a lot, probably based on what we deem to be a reasonable amount of time for a start_upload request to complete (50ms)? Then sample from distribution around 50ms and use this to modify the delay?

Reply:
this isn't what happens empirically, easier to talk about irl